exploration/analysis.py

```diff
@@ -5,6 +5,147 @@
+        abs_skews = [np.abs(s) for s in skews]
+        sum_abs_skews = np.sum(abs_skews)
+
+        # Scale the skewness and KLD by the number of frames
+        scaled_kld = min(kld / block.shape[0], 1.0)
+        scaled_skew = min(sum_abs_skews / block.shape[0], 1.0)
+
+        _labels.loc[_labels['sample_id'] == sid, 'kldiv'] = scaled_kld
+        _labels.loc[_labels['sample_id'] == sid, 'skewness'] = scaled_skew
+
+    return _labels
+
+
+def get_spectra(df: pd.DataFrame, prefix='tx_', band=(0.8, 2.5)):
+    """Generate spectra using Weltch FFT
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        input data that (works better if already cleaned)
+    prefix : str, optional
+        the prefix to look for with each colour, by default 'tx_'
+    band : tuple, optional
+        The bounds of the frequency band within which to measure KL divergence, by default (0.5, 3.5)
+
+    """
+
+    def get_spec(block, field):
+        x = block[field]
+        f, Pxx_spec = welch(x, 30.0, 'flattop',
+                            nperseg=len(x), scaling='spectrum')
+        # only consider certain Hz band
+        spec = Pxx_spec[(f < band[1]) & (f > band[0])]
+        return spec
+
+    colours = ['red', 'green', 'blue']
+    specs = {c: get_spec(df, f"{prefix}{c}") for c in colours}
+
+    return specs
+
+
+def kl_divergence(spectra: dict):
+    """Generate a signal quality score (lower is better) based on mutual KL divergence.
+
+
+    Parameters
+    ----------
+    spectra : dict
+        colour : List[spectra] pairs for each channel
+
+    Returns
+    -------
+    total : float
+        Sum of KL divergence
+    """
+    colours = ['red', 'green', 'blue']
+
+    klds = 0
+    for colour1 in colours:
+        for colour2 in colours:
+            if colour1 != colour2:
+                # Calculate KL-divergence between pairs of colours
+                kld = kl_div(spectra[colour1], spectra[colour2])
+                klds += np.sum(kld)
+    total = np.sum(klds)
+    return total
+
+
+def skewness(df: pd.DataFrame, prefix='mean_', id_field='sample_id') -> pd.DataFrame:
+    """Returns flattened form of the dataframe, with skewness values for each colour channel
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Input data (time series)
+    prefix : str, optional
+        The prefix to look for with each colour, by default 'mean_'
+    id_field : str, optional
+        The field to chunk the data by, by default 'sample_id'
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe with the skewness values.
+    """
+
+    skews = {'sample_id': [], 'skew_red': [],
+             'skew_green': [], 'skew_blue': []}
+
+    for sid in df[id_field].unique():
+        subset = df[df[id_field] == sid]
+        for c in ['red', 'green', 'blue']:
+            skews[f"skew_{c}"].append(scipy.stats.skew(subset[f"{prefix}{c}"]))
+        skews['sample_id'].append(sid)
+
+    return pd.DataFrame(skews)
```

Comment:
Note: this doesn't actually get used by `estimate_quality()`. It's an older version I was dabbling with, but I've left it there just in case.